automata/fa/fa.py

```diff
@@ -2,11 +2,245 @@
+
+        for state in self.iter_states():
+            # every edge needs an origin node, so we add a null node for every
+            # initial state.
+            if self.is_initial(state):
+                # we use a random uuid to make sure that the null node has a
+                # unique id to avoid colliding with other states and null_nodes.
+                null_node = str(uuid.uuid4())
+                graph.node(
+                    null_node,
+                    label="",
+                    shape='point',
+                    fontsize=font_size,
+                )
+                node = self.get_state_name(state)
+                edge_label = "->" + node
+                graph.edge(
+                    null_node,
+                    node,
+                    tooltip=edge_label,
+                    arrowsize=arrow_size,
+                )
+
+        for state in self.iter_states():
+            shape = 'doublecircle' if self.is_accepting(state) else 'circle'
+            node = self.get_state_name(state)
+            graph.node(node, shape=shape, fontsize=font_size)
+
+        if input_str is None:
+            edge_labels = defaultdict(list)
+            for from_state, to_state, symbol in self.iter_transitions():
+                # TODO only do this for NFA
+                label = "ε" if symbol == "" else str(symbol)
+                from_node = self.get_state_name(from_state)
+                to_node = self.get_state_name(to_state)
+                edge_labels[from_node, to_node].append(label)
+
+            for (from_node, to_node), labels in edge_labels.items():
+                label = ",".join(sorted(labels))
+                graph.edge(
+                    from_node,
+                    to_node,
+                    label=label,
+                    arrowsize=arrow_size,
+                    fontsize=font_size,
+                )
+        else:
+            # TODO
+            raise NotImplementedError(
+                "input_str is not yet supported yet"
+            )
+
+            status, taken_transitions_pairs, taken_steps = self.input_check(
+                input_str=input_str,
+                return_result=True
+            )
+            remaining_transitions_pairs = [
+                x
+                for x in all_transitions_pairs
+                if x not in taken_transitions_pairs
+            ]
+
+            start_color = Color("#FFFF00")
+            end_color = Color("#00FF00") if status else Color("#FF0000")
+
+            number_of_colors = len(input_str)
+            interpolation = Color.interpolate([start_color, end_color], space='lch')
+            # TODO why cycle?
+            color_gen = itertools.cycle(
+                interpolation(x / number_of_colors)
+                for x in range(number_of_colors + 1)
+            )
+
+            # Define all transitions in the finite state machine with traversal.
+            counter = 0
+            for pair in taken_transitions_pairs:
```

Comment:
```suggestion
            for from_node, to_node in taken_transitions_pairs:
```

Reply:
Agreed with @eliotwrobson on this, because tuple unpacking is awesome.

Reply:
True, I didn't want to touch this code until I find a way to make the step visualizations work.